File: src/main/java/me/oddlyoko/odoo/modules/reference/OdooManifestReferenceListCondition.java
package me.oddlyoko.odoo.modules.reference;

import com.intellij.patterns.PatternCondition;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.util.ProcessingContext;
import com.jetbrains.python.psi.PyExpression;
import com.jetbrains.python.psi.PyKeyValueExpression;
import com.jetbrains.python.psi.PyListLiteralExpression;
import com.jetbrains.python.psi.PyStringLiteralExpression;
import me.oddlyoko.odoo.modules.OdooModuleUtil;
import org.jetbrains.annotations.NotNull;

import java.util.Set;

public class OdooManifestReferenceListCondition extends PatternCondition<PyStringLiteralExpression> {
    private final Set<String> acceptedKeys;

    public OdooManifestReferenceListCondition(Set<String> acceptedKeys) {
        super("__manifest__");
        this.acceptedKeys = acceptedKeys;
    }

    @Override
    public boolean accepts(@NotNull PyStringLiteralExpression pyString, ProcessingContext context) {
        PsiFile file = pyString.getContainingFile();
        // Only works in manifest file
        if (file == null || !OdooModuleUtil.MANIFEST_FILES.contains(file.getName()))
            return false;

        PsiElement parent = pyString.getParent();
        if (parent instanceof PyListLiteralExpression) {
            parent = parent.getParent();
            if (parent instanceof PyKeyValueExpression) {
                PyExpression key = ((PyKeyValueExpression) parent).getKey();
                if (key instanceof PyStringLiteralExpression)
                    return acceptedKeys.contains(((PyStringLiteralExpression) key).getStringValue());
            }
        }
        return false;
    }
}

File: src/main/kotlin/me/oddlyoko/odoo/modules/reference/OdooManifestReferenceListCondition.kt
package me.oddlyoko.odoo.modules.reference

import com.intellij.patterns.PatternCondition
import com.intellij.util.ProcessingContext
import com.jetbrains.python.psi.PyKeyValueExpression
import com.jetbrains.python.psi.PyListLiteralExpression
import com.jetbrains.python.psi.PyStringLiteralExpression
import me.oddlyoko.odoo.modules.OdooModuleUtil

class OdooManifestReferenceListCondition(private val acceptedKeys: List<String>): PatternCondition<PyStringLiteralExpression>("__manifest__") {

    override fun accepts(pyString: PyStringLiteralExpression, context: ProcessingContext?): Boolean {
        val file = pyString.containingFile ?: return false
        if (!OdooModuleUtil.MANIFEST_FILES.contains(file.name))
            return false

        var parent = pyString.parent
        if (parent is PyListLiteralExpression) {
            parent = parent.parent
            if (parent is PyKeyValueExpression) {
                val key = parent.key
                if (key is PyStringLiteralExpression)
                    return acceptedKeys.contains(key.stringValue)
            }
        }
        return false
    }
}
